cataclysmbnteam/Cataclysm-BN
# Roof buffs

#### Summary



SUMMARY:  [Bugfixes] "Fix missing roofs and partial roof destruction"

#### Purpose of change



Roofs need pretty specific conditions to work correctly:
1. Have `"bash_below": true`
2. Don't have roofs themselves
3. Were spawned on mapgen or load
4. Z-levels are enabled

The first 3 are improved here:
1. A warning is produced when a tile is a roof, but has no `bash_below`. Some terrains had this problems, fixed now.
2. A warning is produced when a roof has a roof. Terrains with this issue fixed.
3. Constructions now create roofs over them if open air was there before.

Additionally, roof bashing is improved:
* Can bash down with `s->SHIFT+.`
* If a tile has its roof wrecked, it itself is wrecked
* If a `bash_below` tile is wrecked, but is not the roof of the tile above, the roof is placed and then bashed with the same force as the original tile. This can lead to 3 destructions (original floor/roof, roof from tile below, the tile below itself).
* Roof and its replacements drop destruction drops and play sounds in correct order (explosions are still reversed)

#### Describe the solution



New checks and changes make it less likely that a roof-less tile with a roof data exists.
Such tiles would get new roofs on reload, so they are in bugged state.

#### Describe alternatives you've considered



#### Testing



New warnings:
* Checkout `data` from current `upload` branch, but with `src` from this PR
* Compile and load the game
* See warnings
* Checkout `data` from this PR
* See no warnings

Constructions:
* Build a log wall
* See it has a flat roof above it

Bashing:
* Set character strength to 80 or so
* Spawn many tiles of laminated glass
* Save+reload to spawn roofs (map editor doesn't spawn them yet)
* Get over one
* Bash down until destroyed
* Repeat a few times
* Notice that in no case an unpassable tile without a roof is created, despite laminated glass being more resistant to damage than flat roof

#### Additional context



Limitations:
* Log walls, when destroyed from above, WILL leave unpassable tiles without roofs
* Unpassable tiles without roofs are not fixed here, only tiles with types that should have roofs but have lost them due to their roofs being bashed
* Many strong tiles have weak roofs, which creates sort of an exploit in that you can bash it from above to destroy it much more easily
* Repeating bash command in down direction is mega clunky
